gossip/bloom.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"encoding/binary"
+	"hash"
+	"sync"
+
+	bloomfilter "github.com/holiman/bloomfilter/v2"
+)
+
+func NewBloomFilterFromBytes(bytes []byte) (*BloomFilter, error) {
+	b, err := NewBloomFilter()
+	if err != nil {
+		return nil, err
+	}
+
+	return b, b.bloomFilter.UnmarshalBinary(bytes)
+}
+
+func NewBloomFilter() (*BloomFilter, error) {
+	return newBloomFilter(1024, 4, 0.75)
+}
+
+func newBloomFilter(m, k uint64, refreshRatio float64) (*BloomFilter, error) {
+	bloomFilter, err := bloomfilter.New(m, k)
+	if err != nil {
+		return nil, err
+	}
+
+	return &BloomFilter{
+		bloomFilter:  bloomFilter,
+		m:            m,
+		k:            k,
+		refreshRatio: refreshRatio,
+	}, nil
+}
+
+// BloomFilter is a bloom filter that is periodically refreshed when it breaches a given
+// threshold of items on Adds
+type BloomFilter struct {
+	m, k         uint64
+	refreshRatio float64
+	bloomFilter  *bloomfilter.Filter
+	lock         sync.RWMutex
+}
+
+// Add invariant: requires that [hash] is at least 8 bytes long.
+func (b *BloomFilter) Add(hash []byte) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.resetIfNeeded()
+
+	binary.BigEndian.Uint64(hash)
+	b.bloomFilter.Add(bloomHasher{Hash: hash})
+}
+
+// assumes [b.lock] is held
+func (b *BloomFilter) resetIfNeeded() {
+	filled := b.bloomFilter.PreciseFilledRatio()
+	if filled >= b.refreshRatio {
+		b.bloomFilter, _ = bloomfilter.New(b.m, b.k)
+	}
+}
```

Comment:
Hmm if we do this here, then this bloom filter implementation seems to lose the main feature of a bloom filter which is probabilistic yes or definitely not set membership.

Ie if you add A, then exceed the cap, A gets wiped, and if you ask if A is there, it may actually tell you no.

Would it make more sense to expose a function that allows the caller to explicitly wipe the bloom filter and then add all of the items that it still wants in the set back to the bloom filter?

Reply:
Yeah i think we can just expose number of elements and caller can wipe as needed.